pydantic/main.py

```diff
@@ -223,6 +237,7 @@ def hash_function(self_: Any) -> int:
 _is_base_model_class_defined = False
 
 
+@__dataclass_transform__(kw_only_default=True, field_descriptors=(Field, FieldInfo))
```

Comment:
Is there any possibility to add the `frozen` argument which should equal `config.frozen or not config.allow_mutation`?

I can't see how this is possible but @erictraut suggested it should be.

Reply:
You can pass frozen=True when constructing a class that derives from `ModelBase`, like this:

```python
class CustomerModel(ModelBase, frozen=True):
```

That was based on a suggestion you made in our email discussion — or at least how I interpreted your suggestion. Did I understand you correctly?

All classes are assumed to be non-frozen unless otherwise specified. There is a request from the `attrs` maintainers to provide a way to specify a `frozen_default` option so they can default to `frozen=True` if it is not specified. If that would also be useful for pydantic, it would bolster the case for adding it to the spec.

Reply:
Oh, that's awesome. I've just tried it and it's already working!

This looks great then, the only problem I can see is the type strictness/laxness which is being discussed on microsoft/pyright#1782

Reply:
I didn't see a way to do it from what I read in the spec.

If I understand correctly, it would be possible for Pyright (and the spec) to understand if the model was created like:

```Python
from pydantic import BaseModel

class Fish(BaseModel, frozen=True):
    title: str
```

...which is currently not supported by pydantic (although I think sounds interesting).

What is currently supported by pydantic (but I understand not by Pyright nor the spec) is:

```Python
from pydantic import BaseModel

class Fish(BaseModel):
    title: str
    class Config:
        frozen = True
```

I'll wait for @erictraut's input in case it's possible to support pydantic's internal `class Config` in a way I didn't realize.

Reply:
It is supported @tiangolo! Config arguments can be passed in class kwargs thanks to @MrMrRobat work in #2356

Reply:
Dang it again! :facepalm: :joy: *pydantic* has evolved so much and I haven't noticed everything! :sweat_smile: 

Thanks for the clarification @PrettyWood, that's awesome! :tada:

Reply:
I tried this with v1.9 and Pyright does not seem to recognise `frozen`:

```py
# pyright: strict

from typing import Hashable


from pydantic import BaseModel

class Foo(BaseModel, frozen=True):
    pass

foo: Hashable = Foo()  # Error: Expression of type "Foo" cannot be assigned to declared type "Hashable"
reveal_type(Foo.__hash__)  # Type of "Foo.__hash__" is "None"


from dataclasses import dataclass

@dataclass(frozen=True)
class Bar:
    pass

bar: Hashable = Bar()  # No error
reveal_type(Bar.__hash__)  # Type of "Bar.__hash__" is "(self: Bar) -> int"

```

Reply:
Interesting, this probably needs a new issue.

I've no idea if this is something pydantic can help with, or an issue with pyright. @erictraut any idea?

Reply:
https://github.com/samuelcolvin/pydantic/issues/3545